validator/src/consensus/keyGen/client.ts

```diff
@@ -0,0 +1,221 @@
+		const verificationShare = createVerificationShare(
+			commitments,
+			participantId,
+		);
+		this.#storage.registerVerification(
+			groupId,
+			groupPublicKey,
+			verificationShare,
+		);
+
+		const coefficients = this.#storage.coefficients(groupId);
+		const participants = this.#storage.participants(groupId);
+		const shares: bigint[] = [];
+		for (const participant of participants) {
+			if (participant.id === participantId) continue;
+			const peerCommitments = commitments.get(participant.id);
+			if (peerCommitments === undefined)
+				throw Error(
+					`Commitments for ${groupId}:${participant.id} are not available!`,
+				);
+			const peerShare = evalPoly(coefficients, participant.id);
+			const encryptedShare = ecdh(
+				peerShare,
+				coefficients[0],
+				peerCommitments[0],
+			);
+			shares.push(encryptedShare);
+		}
+		if (shares.length !== participants.length - 1) {
+			throw Error("Unexpect f length");
+		}
+		await this.#coordinator.publishKeygenSecretShares(
+			groupId,
+			verificationShare,
+			shares,
+		);
+	}
+
+	// `senderIndex` is the index of sending local participant in the participants set
+	// `peerShares` are the calculated and encrypted shares (also defined as `f`)
+	async handleKeygenSecrets(
+		groupId: GroupId,
+		senderId: ParticipantId,
+		peerShares: readonly bigint[],
+	) {
+		const participants = this.#storage.participants(groupId);
+		if (peerShares.length !== participants.length - 1) {
+			throw Error("Unexpect f length");
+		}
+		const participantId = this.#storage.participantId(groupId);
+		if (senderId === participantId) {
+			console.info("Do not handle own share");
+			return;
+		}
+		const commitment = this.#storage.commitments(groupId, senderId);
+		if (commitment === undefined)
+			throw Error(`Commitments for ${groupId}:${senderId} are not available!`);
+		// TODO: check if we should use a reasonable limit for the index (current uint256)
+		const shareIndex =
+			participantId < senderId ? participantId : participantId - 1n;
+		// Note: Number(shareIndex) is theoretically an unsafe cast
+		const key = this.#storage.encryptionKey(groupId);
+		const partialShare = ecdh(
+			peerShares[Number(shareIndex) - 1],
+			key,
+			commitment[0],
+		);
+		const partialVerificationShare = evalCommitment(commitment, participantId);
+		verifyKey(partialVerificationShare, partialShare);
+		this.#storage.registerSecretShare(groupId, senderId, partialShare);
+
+		if (this.#storage.checkIfSecretSharesComplete(groupId)) {
+			const verificationShare = this.#storage.verificationShare(groupId);
+			const secretShares = this.#storage.secretSharesMap(groupId);
+			const signingShare = createSigningShare(secretShares);
+			verifyKey(verificationShare, signingShare);
+			this.#storage.registerSigningShare(groupId, signingShare);
+			this.#storage.clearKeyGen(groupId);
+			const participantId = this.#storage.participantId(groupId);
+			console.info(`Final signing key for ${participantId} calculated`);
+		}
+	}
+}
```

Comment:
General comment, the code is clear and super easy to follow given my FROST knowledge 👏.